File: chad_c/Data_Structures/doubly_linked_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct DoubleNode double_node_t;
typedef struct DoublyLinkedList doubly_linked_list_t;

typedef struct DoubleNode
{
    double_node_t *prev;
    int data;
    double_node_t *next;
} double_node_t;

typedef struct DoublyLinkedList
{
    double_node_t *head;
    double_node_t *tail;
    size_t size;
} doubly_linked_list_t;

doubly_linked_list_t *init_dll()
{
    doubly_linked_list_t *list = malloc(sizeof(doubly_linked_list_t));

    if (list == NULL)
    {
        printf("Failed to allocate memory\n");
        return NULL;
    }

    list->head = NULL;
    list->tail = NULL;
    list->size = 0;

    return list;
}

void free_dll(doubly_linked_list_t *list)
{
    if (list == NULL)
    {
        printf("Bruh\n");
        return;
    }

    double_node_t *curr = list->head;

    while (curr)
    {
        double_node_t *next = curr->next;
        free(curr);
        curr = next;
    }

    free(list);
}

void print_dll(doubly_linked_list_t *list)
{
    if (list == NULL || list->head == NULL || list->size == 0)
    {
        printf("Bruh\n");
        return;
    }

    double_node_t *curr = list->head;

    while (curr)
    {
        if (curr == list->head)
        {
            if (curr->next == NULL)
            {
                printf("NULL <-- [Head: %d] --> NULL\n", curr->data);
                return;
            }

            printf("NULL <--> [Head: %d] <--> ", curr->data);
        }

        else if (curr == list->tail)
        {
            printf("[Tail: %d] <--> NULL\n", curr->data);
            return;
        }

        else
        {
            printf("[%d] <--> ", curr->data);
        }

        curr = curr->next;
    }
}

double_node_t *prepend_dll(doubly_linked_list_t *list, int data)
{
    if (list == NULL)
    {
        printf("Bruh\n");
        return NULL;
    }

    double_node_t *new_node = malloc(sizeof(double_node_t));

    if (new_node == NULL)
    {
        printf("Failed to create new node\n");
        return NULL;
    }

    new_node->prev = NULL;
    new_node->data = data;
    new_node->next = list->head;

    if (list->head)
    {
        list->head->prev = new_node;
    }

    else
    {
        list->tail = new_node;
    }

    list->head = new_node;
    list->size++;

    return new_node;
}

double_node_t *append_dll(doubly_linked_list_t *list, int data)
{
    if (list == NULL)
    {
        printf("Bruh\n");
        return NULL;
    }

    double_node_t *new_node = malloc(sizeof(double_node_t));

    if (new_node == NULL)
    {
        printf("Failed to create new node\n");
        return NULL;
    }

    new_node->prev = list->tail;
    new_node->data = data;
    new_node->next = NULL;

    if (list->tail)
    {
        list->tail->next = new_node;
    }

    else
    {
        list->head = new_node;
    }

    list->tail = new_node;
    list->size++;

    return new_node;
}

double_node_t *insert_node_dll(doubly_linked_list_t *list, int data, size_t i)
{
    if (list == NULL)
    {
        printf("Bruh\n");
        return NULL;
    }

    if (i > list->size)
    {
        printf("Position is out of bound\n");
        return NULL;
    }

    if (i == 0)
    {
        return prepend_dll(list, data);
    }

    if (i == list->size)
    {
        return append_dll(list, data);
    }

    double_node_t *curr = list->head;

    while (i > 1)
    {
        curr = curr->next;
        i--;
    }

    double_node_t *new_node = malloc(sizeof(double_node_t));

    if (new_node == NULL)
    {
        printf("Failed to create new node\n");
        return NULL;
    }

    new_node->prev = curr;
    new_node->data = data;
    new_node->next = curr->next;

    curr->next->prev = new_node;
    curr->next = new_node;

    list->size++;
    return new_node;
}

int remove_node_dll(doubly_linked_list_t *list, int key)
{
    if (list == NULL)
    {
        printf("Bruh\n");
        return -1;
    }

    if (list->head == NULL)
    {
        printf("List is empty\n");
        return -1;
    }

    double_node_t *curr = list->head;

    while (curr)
    {
        if (curr->data == key)
        {
            if (curr == list->head)
            {
                list->head = curr->next;

                if (list->head)
                {
                    list->head->prev = NULL;
                }

                else
                {
                    list->tail = NULL;
                }
            }

            else
            {
                curr->prev->next = curr->next;

                if (curr->next)
                {
                    curr->next->prev = curr->prev;
                }

                else
                {
                    list->tail = curr->prev;
                }
            }

            free(curr);
            list->size--;
            return key;
        }

        curr = curr->next;
    }

    return -1;
}

File: virgin_python/Data_Structures/doubly_linked_list.py
class Node:
    
    def __init__(self, prev = None, data = None, next = None):
        self.prev = prev
        self.data = data
        self.next = next

    def __repr__(self):
        return f"<Node: {self.data}>"


class DoublyLinkedList:
    
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def __len__(self):
        return self.size
    
    def __repr__(self):

        if self.head is None:
            return "[]"
        
        nodes = []
        curr = self.head

        while curr:

            if curr is self.head:
                nodes.append(f"[Head: {curr.data}]")

            elif curr is self.tail:
                nodes.append(f"[Tail: {curr.data}]")

            else:
                nodes.append(f"[{curr.data}]")

            curr = curr.next

        return " <--> ".join(nodes)
    

    def prepend(self, data):
        new_node = Node(None, data, self.head)

        if self.head is not None:
            self.head.prev = new_node
        else:
            self.tail = new_node

        self.head = new_node
        self.size += 1

        return new_node

    def append(self, data):
        new_node = Node(self.tail, data, None)

        if self.tail is not None:
            self.tail.next = new_node
        else:
            self.head = new_node

        self.tail = new_node
        self.size += 1

        return new_node

    def insert(self, data, i):

        if i < 0 or i > self.size:
            raise IndexError("Position is out of bound")

        if i == 0:
            return self.prepend(data)
        
        if i == self.size:
            return self.append(data)


        curr = self.head

        for _ in range(i - 1):
            curr = curr.next

        new_node = Node(curr, data, curr.next)
        curr.next.prev = new_node
        curr.next = new_node

        self.size += 1
        return new_node

    def remove(self, key):

        if self.head is None:
            return None
        
        curr = self.head

        while curr:

            if curr.data == key:

                if curr is self.head:
                    self.head = curr.next

                    if self.head:
                        self.head.prev = None
                    else:
                        self.tail = None

                else:
                    curr.prev.next = curr.next
                    
                    if curr.next:
                        curr.next.prev = curr.prev
                    else:
                        self.tail = curr.prev

                self.size -= 1
                return curr

            curr = curr.next

        return None